### PR TITLE
Use `MacOSRequirement` for casks.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

----

Merge `MaximumMacOSRequirement` into `MacOSRequirement` and also use `MacOSRequirement` for casks.

### CODE REVIEW

File: Library/Homebrew/utils.rb [hunk 1]
```diff
@@ -245,264 +74,437 @@ def inject_dump_stats!(the_module, pattern)
   # rubocop:enable Style/GlobalVars
 end
 
-def with_homebrew_path
-  with_env(PATH: PATH.new(ENV["HOMEBREW_PATH"])) do
-    yield
+module Kernel
+  def require?(path)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils.rb]: Can this refactoring go in another PR?

File: Library/Homebrew/cask/dsl.rb [hunk 2]
```diff
@@ -212,6 +212,7 @@ def depends_on(*args)
       rescue RuntimeError => e
         raise CaskInvalidError.new(cask, e)
       end
+
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/dsl.rb]: Does this refactoring need to be in this PR?

File: Library/Homebrew/requirements/macos_requirement.rb [hunk 3]
```diff
@@ -5,32 +5,58 @@
 class MacOSRequirement < Requirement
   fatal true
 
-  def initialize(tags = [])
-    @version = MacOS::Version.from_symbol(tags.shift) unless tags.empty?
+  def initialize(tags = [], comparator: ">=")
```
Comment (MikeMcQuaid) [inline Library/Homebrew/requirements/macos_requirement.rb]: Thoughts on making `comparator` a symbol like `:maximum` or `:greater_than` instead of a string?

File: Library/Homebrew/utils.rb (hunk 1 above)
Comment (reitermarkus) [reply Library/Homebrew/utils.rb]: Moved to https://github.com/Homebrew/brew/pull/6367.

File: Library/Homebrew/requirements/macos_requirement.rb (hunk 3 above)
Comment (reitermarkus) [reply Library/Homebrew/requirements/macos_requirement.rb]: I think that would just introduce more boilerplate.

Comment (MikeMcQuaid) [reply Library/Homebrew/requirements/macos_requirement.rb]: 🆒
